src/libstore/worker-protocol.hh

```diff
@@ -66,14 +66,104 @@ typedef enum {
 class Store;
 struct Source;
 
-template<class T> T readStorePaths(const Store & store, Source & from);
-
-void writeStorePaths(const Store & store, Sink & out, const StorePathSet & paths);
-
-StorePathCAMap readStorePathCAMap(const Store & store, Source & from);
-
-void writeStorePathCAMap(const Store & store, Sink & out, const StorePathCAMap & paths);
-
-void writeOutputPathMap(const Store & store, Sink & out, const OutputPathMap & paths);
+template<typename T>
+struct WorkerProto {
```

Comment:
What's the motivation for using a templated struct here rather than a namespace like it was done in #3859 (with templated functions inside). Nothing fundamental, but I feel like `worker_proto::read<StorePath>()` is more natural than `WorkerProto<StorePath>::read()`

Reply:
We were worried about implicit conversation so we wanted to "recur" with an explicit `<T>` template instantiation.

However implicit conversations ended up not being a problem, and I've made it so the most important PRs no longer depend on this one so we can go either way.

Reply:
@edolstra What do you think about switching from the workspace to `WorkerProto`? Happy to revert that part if you don't like it. As I wrote above I introduced it because i thought there was a bug. I'm 50-50 on whether to keep it or not.

Reply:
I think I would prefer keeping `worker_proto`. Seeing `WorkerProto` objects constructed everywhere is kind of confusing, since it requires you to read the `WorkerProto` definition to see that they're just a zero-cost wrapper around static methods...

Reply:
Changing it back, but for the record I believe `Class::method`, which is what I'm doing here, syntax never constructs an object of that class?

Reply:
Changed it back! Should be ready now.